lib/SparkPost/SparkPost.php

```diff
@@ -2,151 +2,238 @@
-        $httpConfig->setBaseUri($baseUrl);
-        $httpConfig->setTimeout($this->config['timeout']);
-        $httpConfig->setUserAgent('php-sparkpost/'.$this->version);
-
-        return $httpConfig;
+        $options = $this->options;
+
+        $paramsArray = [];
+        foreach ($params as $key => $value) {
+            if (is_array($value)) {
+                $value = implode(',', $value);
+            }
+
+            array_push($paramsArray, $key.'='.$value);
+        }
+
+        $paramsString = implode('&', $paramsArray);
+
+        return $options['protocol'].'://'.$options['host'].($options['port'] ? ':'.$options['port'] : '').'/api/'.$options['version'].'/'.$path.($paramsString ? '?'.$paramsString : '');
     }
 
-    /**
-     * Validates and sets up the httpAdapter.
-     *
-     * @param $httpAdapter \Ivory\HttpAdapter\HttpAdapterInterface to make requests through.
+    /** 
+     * Sets $httpClient to be used for request.
      *
-     * @throws \Exception
+     * @param Http\Client\HttpClient $httpClient - the client to be used for request
      */
-    public function setHttpAdapter(HttpAdapterInterface $httpAdapter)
+    public function setHttpClient(HttpClient $httpClient)
     {
-        $this->httpAdapter = $httpAdapter;
-        $this->httpAdapter->setConfiguration($this->getHttpConfig($this->config));
+        $this->httpClient = $httpClient;
     }
 
     /**
-     * Allows the user to pass in values to override the defaults and set their API key.
-     *
-     * @param string | array $settingsConfig - Hashmap that contains config values
-     *                                       for the SDK to connect to SparkPost. If its a string we assume that
-     *                                       its just they API Key.
+     * Sets the options from the param and defaults for the SparkPost object.
      *
-     * @throws \Exception
+     * @param array $options - either an string API key or an array of options
      */
-    public function setConfig($settingsConfig)
+    public function setOptions($options)
     {
-        // if the config map is a string we should assume that its an api key
-        if (is_string($settingsConfig)) {
-            $settingsConfig = ['key' => $settingsConfig];
+        // if the options map is a string we should assume that its an api key
+        if (is_string($options)) {
+            $options = ['key' => $options];
         }
 
         // Validate API key because its required
-        if (!isset($settingsConfig['key']) || !preg_match('/\S/', $settingsConfig['key'])) {
+        if (!isset($this->options['key']) && (!isset($options['key']) || !preg_match('/\S/', $options['key']))) {
             throw new \Exception('You must provide an API key');
         }
 
-        $this->config = self::$apiDefaults;
+        $this->options = isset($this->options) ? $this->options : self::$defaultOptions;
```

Comment:
I think we need to use `array_merge` here. The user may not specify all possible options.